text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+```rust
+fn get_work_storage(key: &[u8]) -> Result<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]);
+```
+
+#### Accumulate
+
+The second stage is that of Accumulate. This function is called on-chain only *after* the availability process has completed.
+
+The function signature to the `accumulate` entry-point in the Work Class's code blob is:
+
+```rust
+fn accumulate(results: Vec<(Authorization, Vec<(ItemHash, WorkResult)>)>);
+type ItemHash = [u8; 32];
+```
+
+The logic in `accumulate` may need to know how the various Work Items arrived into a processed Work Package. Since a Work Package could have multiple Work Items of the same Work Class, it makes sense to have a separate inner `Vec` for Work Items sharing the Authorization (by virtue of being in the same Work Package).
+
+Work Items are identified by their Blake2-256 hash, known at the *Item Hash* (`ItemHash`). We provide both the Authorization of the Package and the constituent Work Item Hashes and their Results in order to allow the `refine` logic to take appropriate action in the case that an invalid Work Item was submitted (i.e. one which caused its Refine operation to panic or time-out).
+
+_(Note for later: We may wish to provide a more light-client friendly Work Item identifier than a simple hash; perhaps a Merkle root of equal-size segments.)_
+
+There is an amount of weight which it is allowed to use before being forcibly terminated and any non-committed state changes lost. The lowest amount of weight provided to `accumulate` is defined as the number of `WorkResult` values passed in `results` to `accumulate` multiplied by the `accumulate` field of the Work Class's weight requirements.
+
+However, the actual amount of weight may be substantially more. Each Work Package is allotted a specific amount of weight for all on-chain activity (`weight_per_package` above) and has a weight liability defined by the weight requirements of all Work Items it contains (`total_weight_requirement` above). Any weight remaining after the liability (i.e. `weight_per_package - total_weight_requirement`) may be apportioned to the Work Classes of Items within the Report on a pro-rata basis according to the amount of weight they utilized during `refine`. Any weight unutilized by Classes within one Package may be carried over to the next Package and utilized there.
+
+```rust
+fn get_work_storage(key: &[u8]) -> Option<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]) -> Option<u32>;
+fn checkpoint() -> Weight;
+fn weight_remaining() -> Weight;
+fn set_work_storage(key: &[u8], value: &[u8]) -> Result<(), ()>;
+fn remove_work_storage(key: &[u8]);
+fn ump_enqueue(message: &[u8]) -> Result<(), ()>;
+```
+
+Read-access to the entire Relay-chain state is allowed. No direct write access may be provided since `accumulate` is untrusted code. `set_storage` may fail if an insufficient deposit is held under the Work Class's account.
```

Comment:
```suggestion
Read-access to the entire Relay-chain state is allowed. No direct write access may be provided since `accumulate` is untrusted code. `set_work_storage` may fail if an insufficient deposit is held under the Work Class's account.
```